pkg/logentry/stages/metrics.go

```diff
@@ -0,0 +1,252 @@
+package stages
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+
+	"github.com/grafana/loki/pkg/logentry/metric"
+)
+
+const customPrefix = "promtail_custom_"
+
+const (
+	MetricTypeCounter   = "counter"
+	MetricTypeGauge     = "gauge"
+	MetricTypeHistogram = "histogram"
+
+	ErrEmptyMetricsStageConfig = "empty metric stage configuration"
+)
+
+// MetricConfig is a single metrics configuration.
+type MetricConfig struct {
+	MetricType  string      `mapstructure:"type"`
+	Description string      `mapstructure:"description"`
+	Source      *string     `mapstructure:"source"`
+	Config      interface{} `mapstructure:"config"`
+}
+
+// MetricsConfig is a set of configured metrics.
+type MetricsConfig map[string]MetricConfig
+
+func validateMetricsConfig(cfg MetricsConfig) error {
+	if cfg == nil {
+		return errors.New(ErrEmptyMetricsStageConfig)
+	}
+	for name, config := range cfg {
+		//If the source is not defined, default to the metric name
+		if config.Source == nil {
```

Comment:
why would the source be the name of the metrics ?

Reply:
Just to have a way to define the metric a little more concisely 

```yaml
pipeline_stages:
- json:
    expressions:
      level: log.level
- metric:
    level:
      type: Counter
      description: log level warn counter
      source: level                         ①
      config:
        value: WARN
        action: inc
```
① This can be omitted if your metric name matches the extracted map's key